schulcloud/News+.swift

```diff
@@ -0,0 +1,91 @@
+//
+
+import Foundation
+import BrightFutures
+import CoreData
+import Marshal
+
+// MARK: JSON Processing
+extension News {
+
+    static func upsert(inContext context: NSManagedObjectContext, object: MarshaledObject) -> Future<News, SCError> {
+        
+        do {
+           
+            let fetchRequest = NSFetchRequest<News>(entityName: "News")
+            let id: String = try object.value(for: "_id")
+            fetchRequest.predicate = NSPredicate(format: "id == %@", id)
+            
+            let result = try context.fetch(fetchRequest)
+            if result.count > 1 {
+                throw SCError.database("Found more than one result for \(fetchRequest)")
+            }
+
+            // if news the article is not updated, just return local one
+            if let news = result.first {
+                let storedUpdatedDate = news.updatedAt
+                let updatedDate: NSDate? = try? object.value(for:"updatedAt")
+                
+                // either both nil, or both date are equals
+                if storedUpdatedDate == updatedDate {
+                    return Future(value: news) // the news data is up to date, return it
+                }
+            }
+            
+            // insert / update the news article
+            let news = result.first ?? News(context: context)
+            
+            news.id = id
+            news.title       = try object.value(for: "title")
+            news.content     = try object.value(for: "content")
+            news.displayedAt = try object.value(for: "displayedAt")
+            news.createdAt   = try object.value(for: "createdAt")
+            news.updatedAt   = try? object.value(for: "updatedAt")
+            news.schoolId    = try? object.value(for: "schoolId")
+            news.history     = try object.value(for: "history")
+            
+            let creatorId: String = try object.value(for: "creatorId")
+            let updaterId: String? = try? object.value(for: "updaterId")
+            
+            let creatorFuture = news.fetchCreator(by: creatorId, context: context).onErrorLogAndRecover(with: Void() )
+            let updaterFuture = news.fetchUpdater(by: updaterId, context: context).onErrorLogAndRecover(with: Void() )
+            
+            return [creatorFuture, updaterFuture].sequence().flatMap(object: news)
+        } catch let error as MarshalError {
+            return Future(error: .jsonDeserialization(error.description))
+        } catch let error {
+            return Future(error: .database(error.localizedDescription))
+        }
+    }
+    
+    func fetchCreator(by id: String, context: NSManagedObjectContext) -> Future< Void, SCError > {
+        return User.fetchQueue.sync {
+            return User.fetch(by: id, inContext:context).flatMap { creator -> Future<Void, SCError> in
+                self.creator = creator
+                return Future(value: Void() )
+            }
```

Comment:
If we only handle the case when this user request is successful, shouldn't the `creator` attribute on `News` be optional?

Reply:
If I understand well, the request might fail which means we won't have a creator to assign. Since it isn't important, we are ok with not having a creator and we can make the property optional. Do I understand you correctly?

Reply:
Correct